things/lift/modbus/js/main.js

```diff
@@ -0,0 +1,179 @@
+const { parseArgs } = require('node:util')
+const express = require('express')
+require('dotenv').config()
+
+const app = express()
+
+const thingName = "modbus-lift"
+const hostname = "0.0.0.0"
+let portNumber = "8502"
+const thingUnitID = 1
+
+const { values: { port } } = parseArgs({
+    options: {
+        port: {
+            type: 'string',
+            short: 'p'
+        }
+    }
+})
+  
+if (port && !isNaN(parseInt(port))) {
+    portNumber = parseInt(port)
+}
+
+const tmPath = process.env.TM_PATH
+
+if (process.platform === 'win32') {
+    tmPath.split(path.sep).join(path.win32.sep)
+}
+
+const thingModel = JSON.parse(fs.readFileSync(path.join(__dirname, tmPath)))
+
+const placeholderReplacer = new JsonPlaceholderReplacer()
+placeholderReplacer.addVariableMap({
+    PROTOCOL: 'modbus+tcp',
+    THING_NAME: thingName,
+    HOSTNAME: hostname,
+    PORT_NUMBER: portNumber
+})
+const thingDescription = placeholderReplacer.replace(thingModel)
+thingDescription['@type'] = 'Thing'
+
+const coils = new Array(9999)
+const discreteInputs = new Array(9999)
+const inputRegisters = new Array(9999)
+const holdingRegisters = new Array(9999)
+
+const lightSwitchForms = [{
+    "href": `${thingUnitID}?address=1&quantity=1`,
+    "op": [
+        "readproperty"
+    ],
+    "modbus:entity": "Coil",
+    "modbus:function": "readCoil",
+    "contentType": "application/octet-stream"
+}, {
+    "href": `${thingUnitID}?address=1&quantity=1`,
+    "op": [
+        "writeproperty"
+    ],
+    "modbus:entity": "Coil",
+    "modbus:function": "writeSingleCoil",
+    "contentType": "application/octet-stream"
+}]
+
+
+thingDescription['properties']['lightSwitch']['forms'] = lightSwitchForms
+
+const onTheMoveForms = [{
+    "href": `${thingUnitID}?address=1&quantity=1`,
+    "op": [
+        "readproperty"
+    ],
+    "modbus:entity": "DiscreteInput",
+    "modbus:function": "readDiscreteInput",
+    "contentType": "application/octet-stream"
+}]
+
+discreteInputs[0] = 0
+thingDescription['properties']['onTheMove']['forms'] = onTheMoveForms
+
+const floorNumberForms = [{
+    "href": `${thingUnitID}?address=1&quantity=1`,
+    "op": [
+        "readproperty"
+    ],
+    "modbus:entity": "HoldingRegister",
+    "modbus:function": "readHoldingRegisters",
+    "contentType": "application/octet-stream"
+}, {
+    "href": `${thingUnitID}?address=1&quantity=1`,
+    "op": [
+        "writeproperty"
+    ],
+    "modbus:entity": "HoldingRegister",
+    "modbus:function": "writeSingleHoldingRegister",
+    "contentType": "application/octet-stream"
+}]
+
+thingDescription['properties']['floorNumber']['forms'] = floorNumberForms
+
+const vector = {
+    getDiscreteInput: function(addr, unitID) {
+        if (thingUnitID === unitID) {
+            console.log(`Reading discrete input @${addr}`)
+            return discreteInputs[addr - 1];
+        }
+    },
+    getInputRegister: function(addr, unitID) {
```

Comment:
can be removed since it is not used

Reply:
Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)